src/Shared/Internal/ReEntrantScopeLock.cs

```diff
@@ -0,0 +1,103 @@
+using System;
+#if !NET35
+using System.Threading;
+#endif
+#if ASP_NET_CORE
+using HttpContextBase = Microsoft.AspNetCore.Http.HttpContext;
+#else
+using System.Web;
+#endif
+
+namespace NLog.Web.Internal
+{
+    /// <summary>
+    /// Manages if a LayoutRenderer can be called recursively.
+    /// Especially used by AspNetSessionValueLayoutRenderer
+    /// 
+    /// Since NET 35 does not support AsyncLocal or even ThreadLocal
+    /// a different technique must be used for that platform
+    /// </summary>
+    internal readonly struct ReEntrantScopeLock : IDisposable
+    {
+        internal ReEntrantScopeLock(HttpContextBase context)
+        {
+            _httpContext = context;
+
+            // The line below is required, because this is a struct, otherwise we get CS0188 compiler error
+            // which is 'The 'this' object cannot be used before all of its fields have been assigned.'
+            _obtainedLock = false;
```

Comment:
Can this assignment be removed, along with the comment above?

Reply:
working on it...

Reply:
If we do that we get this error:
[https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs0188?f1url=%3FappId%3Droslyn%26k%3Dk(CS0188)](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs0188?f1url=%3FappId%3Droslyn%26k%3Dk(CS0188))

If we make it a class, then that error is eliminated.

Reply:
Yes one must assign all fields of a struct in the constructor, but right now you are assigning `_obtainedLock` two times. I would like to have just one assignment (The one from `TryGetLock`)

Reply:
Changed as requested. But, seems .NET does not consider "variable = Method()' as a valid assignment.  The build error in AppVeyor matches the error i see in the VS 2022 output window.